source/particle/world.cc

```diff
@@ -904,15 +1083,14 @@ namespace aspect
         solution[component_indices.pressure] = pressure->get_value(evaluation_point);
         solution[component_indices.temperature] = temperature.get_value(evaluation_point);
 
-        if (n_compositional_fields > 0)
-          for (unsigned int j=0; j<n_compositional_fields; ++j)
-            solution[component_indices.compositional_fields[j]] = get_value<n_compositional_fields>(
-                                                                    compositions.get_value(evaluation_point),
-                                                                    j);
+        for (const auto &eval : compositions)
+          {
+            const std::vector<double> values = eval->get_value(evaluation_point);
```

Comment:
Can we avoid the copy and maybe use a type that avoids the memory allocation altogether (e.g. `boost::container::small_vector`)?

Reply:
This should not allocate (return value optimization, as we return a copy from a locally constructed temporary in that function). I think a bigger optimization would be to do all quadrature points = particles at once. Is there a reason you evaluate them separately?

Reply:
(but we obviously allocate the std::vector inside that function)